src/libutil/xxhash.cpp

```diff
@@ -944,5 +945,49 @@ unsigned long long XXH64_digest (const XXH64_state_t* state_in)
 }
 
 
+#define HSTATE(SZ) ((XXH ## SZ ## _state_t*)m_state)
+
+#if defined(_MSC_VER) || defined(__CYGWIN__)
+  // Windows needs these exported !
+  #define HASH_INLINE_DEFINITION(TYPE, SZ) \
+  template <> OIIO_FORCEINLINE void Hash<TYPE>::append(const std::string &str) { \
+     return append(str.c_str(), str.size()); \
+  } \
+  template <> OIIO_FORCEINLINE void Hash<TYPE>::append(const char *str) { \
+     return append(str, strlen(str)); \
+  }
+#else
+  // Otherwise make the linker error if they were not actually inlined.
+  #define HASH_INLINE_DEFINITION(TYPE, SZ)
+#endif
+
+#define HASH_DEFINITION(TYPE, SZ) \
+  template <> void Hash<TYPE>::reset (TYPE seed) { \
+      XXH ## SZ ## _reset(HSTATE(SZ), seed); \
+  } \
+  template <> Hash<TYPE>::Hash(TYPE seed) : m_state(XXH ## SZ ## _createState()) { \
+      reset(seed); \
+  } \
+  template <> Hash<TYPE>::~Hash() { \
+      XXH ## SZ ## _freeState(HSTATE(SZ)); \
+  } \
+  template <> void Hash<TYPE>::append (const void *val, size_t N) { \
+      m_failed = (XXH ## SZ ## _update(HSTATE(SZ), val, N) != XXH_OK) || m_failed; \
+  } \
+  template <>  TYPE Hash<TYPE>::value () const { \
+      return XXH ## SZ ## _digest(HSTATE(SZ)); \
+  } \
+  template <> void Hash<TYPE>::append (const ustring &str) { \
+      append(str.hash()); \
```

Comment:
I'm thinking that maybe you should drop the `ustring` specialization and just let it use string_view to catch ustrings. I think it could be helpful to ensure that adding a sequence of characters will yield the same hash behavior regardless of whether it's a std::string, C++ string, or ustring. If ustring gives a different hash than the others (because you're directly using the ustring table's internal hash, rather than the xxhash as would be used for other character sequences), it would be different and I can imagine this leading to subtle bugs.